File: ios/Commonplace/Store.swift
import Foundation
import Observation

@Observable
final class Store {

    // MARK: Corpus

    var bookmarks: [Bookmark] = []
    var briefs: [Brief] = []
    var relations: [Int: [Relation]] = [:]
    var askCorpus: [AskAnswer] = []

    // MARK: UI state

    var query = ""
    var filter: Filter = .all
    var activeTag: String?
    var groupBy: GroupBy = .date
    var readBriefs: Set<Int> = [141, 140, 139, 138]
    var settings = AppSettings()
    var toast: String?

    // MARK: Display constants — the prototype's theatre, kept deliberately
    //
    // This is a portfolio build: it has to look like a product in real use, so
    // these are NOT computed from `bookmarks.count`. The web app does the
    // opposite and shows real counts. See 02-interactions-and-state.md § 2.

    let totalIdeas = "1,284"
    let briefCount = 142
    let archiveCount = 143
    let unreviewedCount = 89
    let connectionsTotal = "1,284"
    let syncedAgo = "6 min"
    let archiveSince = "4 March"

    private var toastTask: Task<Void, Never>?

    // MARK: Load

    init() { load() }

    private func load() {
        guard
            let url = Bundle.main.url(forResource: "Seed", withExtension: "json"),
            let data = try? Data(contentsOf: url),
            let seed = try? JSONDecoder().decode(SeedFile.self, from: data)
        else {
            assertionFailure("Seed.json is missing from the bundle")
            return
        }

        bookmarks = seed.bookmarks
        briefs = seed.briefs
        askCorpus = seed.askCorpus

        var graph: [Int: [Relation]] = [:]
        for (key, rows) in seed.relations {
            guard let from = Int(key) else { continue }
            graph[from] = rows.compactMap { row in
                guard let id = row.first?.intValue, let reason = row.last?.stringValue else { return nil }
                // A reason of "Related" or "Similar" is a failure — drop the row.
                guard !["related", "similar"].contains(reason.lowercased()) else { return nil }
                return Relation(id: id, reason: reason)
            }
        }
        relations = graph
    }

    // MARK: Lookup
    //
    // Briefs, connections and Ask answers all reference bookmarks BY ID and
    // hydrate at read time. Nothing is denormalised, so marking something
    // reviewed in the library updates it inside the brief too.

    func bookmark(_ id: Int) -> Bookmark? { bookmarks.first { $0.id == id } }

    func hydrate(_ ids: [Int]) -> [Bookmark] { ids.compactMap(bookmark) }

    func related(_ id: Int) -> [(relation: Relation, bookmark: Bookmark)] {
        (relations[id] ?? []).compactMap { r in
            guard let b = bookmark(r.id) else { return nil }
            return (r, b)
        }
    }

    // MARK: Mutations

    func toggleReviewed(_ id: Int) {
        guard let i = bookmarks.firstIndex(where: { $0.id == id }) else { return }
        bookmarks[i].reviewed.toggle()
        flash(bookmarks[i].reviewed ? "Marked reviewed" : "Moved back to unreviewed")
    }

    func toggleCore(_ id: Int) {
        guard let i = bookmarks.firstIndex(where: { $0.id == id }) else { return }
        bookmarks[i].core.toggle()
        flash(bookmarks[i].core ? "Added to core — resurfaces more often" : "Removed from core")
    }

    func setNote(_ id: Int, _ note: String) {
        guard let i = bookmarks.firstIndex(where: { $0.id == id }) else { return }
        bookmarks[i].note = note
    }

    func addTag(_ id: Int, _ tag: String) {
        let t = tag.trimmingCharacters(in: .whitespaces)
        guard !t.isEmpty, let i = bookmarks.firstIndex(where: { $0.id == id }),
              !bookmarks[i].tags.contains(t) else { return }
        bookmarks[i].tags.append(t)
    }

    func removeTag(_ id: Int, _ tag: String) {
        guard let i = bookmarks.firstIndex(where: { $0.id == id }) else { return }
        bookmarks[i].tags.removeAll { $0 == tag }
    }

    func toggleBriefRead(_ no: Int) {
        if readBriefs.contains(no) {
            readBriefs.remove(no)
        } else {
            readBriefs.insert(no)
            // Toasts only when marking read, not when unmarking.
            flash("Brief marked read · next one at \(settings.time)")
        }
    }

    func isRead(_ no: Int) -> Bool { readBriefs.contains(no) }

    /// The toast auto-dismisses after 1.9s.
    func flash(_ message: String) {
        toastTask?.cancel()
        toast = message
        toastTask = Task { @MainActor [weak self] in
            try? await Task.sleep(for: .milliseconds(1900))
            guard !Task.isCancelled else { return }
            self?.toast = nil
        }
    }

    // MARK: Derived

    /// Filtered by filter, then activeTag, then a case-insensitive substring
    /// match of query against title + summary + author + tags.
    var visible: [Bookmark] {
        let q = query.trimmingCharacters(in: .whitespaces).lowercased()
        return bookmarks.filter { b in
            if filter == .unreviewed && b.reviewed { return false }
            if filter == .core && !b.core { return false }
            if let tag = activeTag, !b.tags.contains(tag) { return false }
            guard !q.isEmpty else { return true }
            let hay = "\(b.title) \(b.summary) \(b.author) \(b.tags.joined(separator: " "))"
            return hay.lowercased().contains(q)
        }
    }

    struct Bucket: Identifiable {
        var id: String { label }
        let label: String
        let items: [Bookmark]
    }

    /// Theme buckets sort by descending size; `.none` yields one unlabelled bucket.
    var buckets: [Bucket] {
        let items = visible
        guard groupBy != .none else {
            return items.isEmpty ? [] : [Bucket(label: "", items: items)]
        }

        var order: [String] = []
        var grouped: [String: [Bookmark]] = [:]
        for b in items {
            let key = groupBy == .date ? Self.monthGroup(b.date) : Palette.theme(for: b.tags)
            if grouped[key] == nil { order.append(key) }
            grouped[key, default: []].append(b)
        }

        let result = order.map { Bucket(label: $0, items: grouped[$0] ?? []) }
        return groupBy == .theme ? result.sorted { $0.items.count > $1.items.count } : result
    }

    /// Tags with counts, alphabetical — the desktop sidebar's tag list becomes a
    /// horizontally scrolling row above the library.
    var tagList: [(name: String, count: Int)] {
        var counts: [String: Int] = [:]
        for b in bookmarks { for t in b.tags { counts[t, default: 0] += 1 } }
        // Counts keep the prototype's scaling so the library looks lived-in.
        return counts.keys.sorted().map { ($0, counts[$0]! * 17 + 6) }
    }

    private static let months = [
        "Jan": "January", "Feb": "February", "Mar": "March", "Apr": "April",
        "May": "May", "Jun": "June", "Jul": "July", "Aug": "August",
        "Sep": "September", "Oct": "October", "Nov": "November", "Dec": "December",
    ]

    static func monthGroup(_ date: String) -> String {
        let parts = date.split(separator: " ")
        let month = parts.count > 1 ? String(parts[1]) : ""
        return "\(months[month] ?? month) 2026"
    }

    // MARK: Ask
    //
    // There is no error state and no empty-result illustration. A miss is framed
    // as information about the library rather than about the question.

    func answer(for question: String) -> AskAnswer {
        let q = question.lowercased()

        if let hit = askCorpus.first(where: { $0.keys.contains { q.contains($0) } }) {
            return hit
        }

        let needle = q.split(separator: " ").first { $0.count > 3 }.map(String.init) ?? q
        let matches = bookmarks.filter {
            "\($0.title) \($0.summary) \($0.tags.joined(separator: " "))"
                .lowercased().contains(needle)
        }.prefix(3)

        if !matches.isEmpty {
            return AskAnswer(
                keys: [],
                label: "Across \(matches.count) loosely related saves",
                paras: [
                    "Nothing in your library answers that directly. The closest saves are below — they touch the question without settling it.",
                    "If this is a live question for you, it is worth saving deliberately rather than waiting for it to turn up.",
                ],
                ids: matches.map(\.id)
            )
        }

        return AskAnswer(
            keys: [],
            label: "No matches",
            paras: ["Nothing you have saved speaks to that. That is useful information about the library rather than about the question."],
            ids: []
        )
    }

    static let askSuggestions = [
        "What have I saved about pricing strategy?",
        "What do my saves say about building in public?",
        "Everything on sleep and energy",
        "How do I decide faster?",
    ]
}

File: android/app/src/main/java/com/commonplace/Store.kt
package com.commonplace

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.platform.LocalContext
import com.commonplace.design.Palette
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.int

/**
 * All state and all mutations — a port of ios/Commonplace/Store.swift.
 *
 * Briefs, connections and Ask answers all reference bookmarks BY ID and hydrate
 * at read time. Nothing is denormalised, so marking something reviewed in the
 * library updates it inside the brief too. Keep it that way.
 */
class Store(seed: SeedFile, private val scope: CoroutineScope) {

    // ── Corpus ─────────────────────────────────────────────────────────────

    var bookmarks by mutableStateOf(seed.bookmarks)
        private set

    val briefs: List<Brief> = seed.briefs
    val askCorpus: List<AskAnswer> = seed.askCorpus

    /**
     * A reason of "Related" or "Similar" is a failure — those rows are dropped
     * on load rather than rendered.
     */
    val relations: Map<Int, List<Relation>> = seed.relations.mapNotNull { (key, rows) ->
        val from = key.toIntOrNull() ?: return@mapNotNull null
        from to rows.mapNotNull { row ->
            val id = row.getOrNull(0)?.int ?: return@mapNotNull null
            val reason = row.getOrNull(1)?.content ?: return@mapNotNull null
            if (reason.lowercase() in setOf("related", "similar")) null
            else Relation(id, reason)
        }
    }.toMap()

    // ── UI state ───────────────────────────────────────────────────────────

    var query by mutableStateOf("")
    var filter by mutableStateOf(Filter.All)
    var activeTag by mutableStateOf<String?>(null)
    var groupBy by mutableStateOf(GroupBy.Date)
    var settings by mutableStateOf(AppSettings())
    var toast by mutableStateOf<String?>(null)
        private set

    private var readBriefs by mutableStateOf(setOf(141, 140, 139, 138))

    // ── Display constants — the prototype's theatre, kept deliberately ─────
    //
    // This is a portfolio build: it has to look like a product in real use, so
    // these are NOT computed from `bookmarks.size`. The web app does the
    // opposite and shows real counts. See 02-interactions-and-state.md § 2.

    val totalIdeas = "1,284"
    val briefCount = 142
    val archiveCount = 143
    val unreviewedCount = 89
    val connectionsTotal = "1,284"
    val syncedAgo = "6 min"
    val archiveSince = "4 March"

    private var toastJob: Job? = null

    // ── Lookup ─────────────────────────────────────────────────────────────

    fun bookmark(id: Int): Bookmark? = bookmarks.firstOrNull { it.id == id }

    fun hydrate(ids: List<Int>): List<Bookmark> = ids.mapNotNull { bookmark(it) }

    /**
     * The detail screen's ↑ ↓ buttons step through the WHOLE corpus with
     * wraparound, not the filtered subset — reading is browsing, and filters
     * are for finding.
     */
    fun step(id: Int, direction: Int): Int {
        if (bookmarks.isEmpty()) return id
        val i = bookmarks.indexOfFirst { it.id == id }
        if (i < 0) return id
        val n = bookmarks.size
        return bookmarks[((i + direction) % n + n) % n].id
    }

    fun related(id: Int): List<Pair<Relation, Bookmark>> =
        (relations[id] ?: emptyList()).mapNotNull { r ->
            bookmark(r.id)?.let { r to it }
        }

    // ── Mutations ──────────────────────────────────────────────────────────

    private fun update(id: Int, transform: (Bookmark) -> Bookmark) {
        bookmarks = bookmarks.map { if (it.id == id) transform(it) else it }
    }

    fun toggleReviewed(id: Int) {
        val next = bookmark(id)?.reviewed?.not() ?: return
        update(id) { it.copy(reviewed = next) }
        flash(if (next) "Marked reviewed" else "Moved back to unreviewed")
    }

    fun toggleCore(id: Int) {
        val next = bookmark(id)?.core?.not() ?: return
        update(id) { it.copy(core = next) }
        flash(if (next) "Added to core — resurfaces more often" else "Removed from core")
    }

    fun setNote(id: Int, note: String) = update(id) { it.copy(note = note) }

    fun addTag(id: Int, tag: String) {
        val t = tag.trim()
        if (t.isEmpty()) return
        val current = bookmark(id) ?: return
        if (current.tags.contains(t)) return
        update(id) { it.copy(tags = it.tags + t) }
    }

    fun removeTag(id: Int, tag: String) = update(id) { it.copy(tags = it.tags - tag) }

    fun toggleBriefRead(no: Int) {
        readBriefs = if (no in readBriefs) {
            readBriefs - no
        } else {
            // Toasts only when marking read, not when unmarking.
            flash("Brief marked read · next one at ${settings.time}")
            readBriefs + no
        }
    }

    fun isRead(no: Int): Boolean = no in readBriefs

    /** The toast auto-dismisses after 1.9s. */
    fun flash(message: String) {
        toastJob?.cancel()
        toast = message
        toastJob = scope.launch {
            delay(1900)
            toast = null
        }
    }

    // ── Derived ────────────────────────────────────────────────────────────

    /**
     * Filtered by filter, then activeTag, then a case-insensitive substring
     * match of query against title + summary + author + tags.
     */
    val visible: List<Bookmark> by derivedStateOf {
        val q = query.trim().lowercase()
        bookmarks.filter { b ->
            when {
                filter == Filter.Unreviewed && b.reviewed -> false
                filter == Filter.Core && !b.core -> false
                activeTag != null && !b.tags.contains(activeTag) -> false
                q.isEmpty() -> true
                else -> "${b.title} ${b.summary} ${b.author} ${b.tags.joinToString(" ")}"
                    .lowercase().contains(q)
            }
        }
    }

    data class Bucket(val label: String, val items: List<Bookmark>)

    /** Theme buckets sort by descending size; `None` yields one unlabelled bucket. */
    val buckets: List<Bucket> by derivedStateOf {
        val items = visible
        if (groupBy == GroupBy.None) {
            if (items.isEmpty()) emptyList() else listOf(Bucket("", items))
        } else {
            val grouped = LinkedHashMap<String, MutableList<Bookmark>>()
            for (b in items) {
                val key = if (groupBy == GroupBy.Date) monthGroup(b.date) else Palette.theme(b.tags)
                grouped.getOrPut(key) { mutableListOf() }.add(b)
            }
            val result = grouped.map { (label, items) -> Bucket(label, items) }
            if (groupBy == GroupBy.Theme) result.sortedByDescending { it.items.size } else result
        }
    }

    data class TagCount(val name: String, val count: Int)

    /**
     * Tags with counts, alphabetical — the desktop sidebar's tag list becomes a
     * horizontally scrolling row above the library.
     */
    val tagList: List<TagCount> by derivedStateOf {
        val counts = mutableMapOf<String, Int>()
        for (b in bookmarks) for (t in b.tags) counts[t] = (counts[t] ?: 0) + 1
        // Counts keep the prototype's scaling so the library looks lived-in.
        counts.keys.sorted().map { TagCount(it, counts.getValue(it) * 17 + 6) }
    }

    // ── Ask ────────────────────────────────────────────────────────────────
    //
    // There is no error state and no empty-result illustration. A miss is framed
    // as information about the library rather than about the question.

    fun answer(question: String): AskAnswer {
        val q = question.lowercase()

        askCorpus.firstOrNull { a -> a.keys.any { q.contains(it) } }?.let { return it }

        val needle = q.split(" ").firstOrNull { it.length > 3 } ?: q
        val matches = bookmarks.filter {
            "${it.title} ${it.summary} ${it.tags.joinToString(" ")}"
                .lowercase().contains(needle)
        }.take(3)

        if (matches.isNotEmpty()) {
            return AskAnswer(
                keys = emptyList(),
                label = "Across ${matches.size} loosely related saves",
                paras = listOf(
                    "Nothing in your library answers that directly. The closest saves are below — they touch the question without settling it.",
                    "If this is a live question for you, it is worth saving deliberately rather than waiting for it to turn up.",
                ),
                ids = matches.map { it.id },
            )
        }

        return AskAnswer(
            keys = emptyList(),
            label = "No matches",
            paras = listOf("Nothing you have saved speaks to that. That is useful information about the library rather than about the question."),
            ids = emptyList(),
        )
    }

    companion object {
        private val months = mapOf(
            "Jan" to "January", "Feb" to "February", "Mar" to "March", "Apr" to "April",
            "May" to "May", "Jun" to "June", "Jul" to "July", "Aug" to "August",
            "Sep" to "September", "Oct" to "October", "Nov" to "November", "Dec" to "December",
        )

        fun monthGroup(date: String): String {
            val parts = date.split(" ")
            val month = if (parts.size > 1) parts[1] else ""
            return "${months[month] ?: month} 2026"
        }

        val askSuggestions = listOf(
            "What have I saved about pricing strategy?",
            "What do my saves say about building in public?",
            "Everything on sleep and energy",
            "How do I decide faster?",
        )

        private val json = Json { ignoreUnknownKeys = true }

        /** Seed.json is bundled from ios/Commonplace/Resources — see app/build.gradle.kts. */
        fun loadSeed(context: Context): SeedFile =
            context.assets.open("Seed.json").bufferedReader().use {
                json.decodeFromString(SeedFile.serializer(), it.readText())
            }
    }
}

val LocalStore = staticCompositionLocalOf<Store> { error("No Store provided") }

@Composable
fun rememberStore(): Store {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    return remember { Store(Store.loadSeed(context), scope) }
}
